src/xdsl/parser.py

```diff
@@ -377,6 +377,26 @@ def parse_optional_string(self,
 
     T = TypeVar('T')
 
+    def parse_nested_list(self,
+                          parse_optional_one: Callable[[], T | None],
+                          delimiter: str = ",",
+                          skip_white_space: bool = True) -> list[T]:
+        '''
+        Parse and flatten a list of lists. The result is a list of elements, no matter the
+        rank of the input.
+        '''
+        if self.parse_optional_char("["):
+            result = [
+                el for els in self.parse_list(lambda: self.parse_nested_list(
+                    parse_optional_one, delimiter, skip_white_space))
+                for el in els
+            ]
+            self.parse_char("]")
+        else:
+            result = self.parse_list(parse_optional_one, delimiter,
```

Comment:
Shouldn't this always return a single value? Is there a case you may parse a list without the `[` and `]` delimiters?

Reply:
Yeah, we always parse what's on the inside of the `[` and `]` delimiters, this function just parses comma-separated values.

Reply:
Another way of doing that is calling `self.peek_char("[")` to know if we have to parse nested lists, or just a list of values.